forest_benchmarking/quantum_volume.py

```diff
@@ -0,0 +1,262 @@
+
+    :param bit_array: an array of bits with right-most bit considered least significant.
+    :return: the integer corresponding to the bitstring.
+    """
+    output = 0
+    for bit in bit_array:
+        output = (output << 1) | bit
+    return output
+
+
+def _naive_program_generator(qc: QuantumComputer, qubits: Sequence[int], permutations: np.ndarray,
+                             gates: np.ndarray) -> Program:
+    """
+    Naively generates a native quil program to implement the circuit which is comprised of the given
+    permutations and gates.
+
+    :param qc: the quantum resource that will implement the PyQuil program for each model circuit
+    :param qubits: the qubits available for the implementation of the circuit. This naive
+        implementation simply takes the first depth-many available qubits.
+    :param permutations: array of depth-many arrays of size n_qubits indicating a qubit permutation
+    :param gates: a depth by depth//2 array of matrices representing the 2q gates at each layer.
+        The first row of matrices is the earliest-time layer of 2q gates applied.
+    :return: a PyQuil program in native_quil instructions that implements the circuit represented by
+        the input permutations and gates. The list of qubits specifies the qubits from which to
+        read the output of the program. The qubits are listed in appropriate order for comparison
+        with simulated heavy hitters from collect_heavy_outputs.
+    """
+    num_measure_qubits = len(permutations[0])
+    # at present, naively select the minimum number of qubits with smallest labels to run on
+    qubits = sorted(qubits)[:num_measure_qubits]
+
+    # create a simple program that uses the compiler to directly generate 2q gates from the matrices
+    prog = Program()
+    for layer_idx, (perm, layer) in enumerate(zip(permutations, gates)):
+        for gate_idx, gate in enumerate(layer):
+            # get the Quil definition for the new gate
+            g_definition = DefGate("LYR" + str(layer_idx) + "_RAND" + str(gate_idx), gate)
+            # get the gate constructor
+            G = g_definition.get_constructor()
+            # add definition to program
+            prog += g_definition
+            # add gate to program, acting on properly permuted qubits
+            prog += G(int(qubits[perm[gate_idx]]), int(qubits[perm[gate_idx+1]]))
+
+    ro = prog.declare("ro", "BIT", len(qubits))
+    for idx, qubit in enumerate(qubits):
+        prog.measure(qubit, ro[idx])
+
+    native_quil = qc.compiler.quil_to_native_quil(prog)
+
+    return native_quil
+
+
+def collect_heavy_outputs(wfn_sim: NumpyWavefunctionSimulator, permutations: np.ndarray,
+                          gates: np.ndarray) -> List[int]:
+    """
+    Uses the provided wfn_sim to calculate the probability of measuring each bitstring from the
+    output of the circuit comprised of the given permutations and gates; those 'heavy' bitstrings
```

Comment:
Split after the `;` for a shorter one(-ish) first line summary

Reply:
Sorry for harping on this. When we render API docs in sphinx it becomes important. The first line will go in the table of functions and the rest of it will be rendered nicely when you click through